File: src/types/addon.ts
/** Represents an addon installed on the local system */
export interface InstalledAddon {
  id: number;
  slug: string;
  name: string;
  installedVersion: string;
  sourceType: 'index' | 'github' | 'local';
  sourceRepo?: string;
  installedAt: string;
  updatedAt: string;
  autoUpdate: boolean;
  manifestPath: string;
  /** Pre-computed sort key from index for version comparison (index addons only) */
  versionSortKey?: number;
  /** Commit SHA for branch-based version tracking */
  commitSha?: string;
}

/** Version tracking info for simplified update detection */
export interface VersionTracking {
  /** Pre-computed sort key from index for direct integer comparison */
  versionSortKey?: number;
  /** Commit SHA for branch-based version tracking */
  commitSha?: string;
}

/** Parsed addon manifest from .txt file */
export interface AddonManifest {
  title: string;
  apiVersion?: string;
  author?: string;
  version?: string;
  description?: string;
  dependencies: string[];
  optionalDependencies: string[];
  savedVariables: string[];
  files: string[];
}

/** Information about an available update */
export interface UpdateInfo {
  slug: string;
  name: string;
  currentVersion: string;
  newVersion: string;
  downloadUrl: string;
  sourceType: 'index' | 'github' | 'local';
  sourceRepo?: string;
  installInfo?: import('./index').InstallInfo;
  /** Multiple download sources with fallback (jsDelivr CDN -> GitHub archive) */
  downloadSources?: import('./index').DownloadSource[];
}

/** Download progress event */
export interface DownloadProgress {
  slug: string;
  status: 'pending' | 'downloading' | 'extracting' | 'complete' | 'failed';
  progress: number;
  error?: string;
}

/** Locally scanned addon info */
export interface ScannedAddon {
  name: string;
  path: string;
  manifest: AddonManifest;
  hasSavedVariables: boolean;
}

/** A resolved dependency ready for installation */
export interface ResolvedDependency {
  /** The addon slug */
  slug: string;
  /** Display name */
  name: string;
  /** Version to install */
  version: string;
  /** Download URL */
  downloadUrl: string;
  /** Installation info from the index */
  installInfo: import('./index').InstallInfo;
  /** Depth in the dependency tree (0 = direct dependency) */
  depth: number;
}

/** Result of dependency resolution */
export interface DependencyResult {
  /** Dependencies that can be installed from the index */
  resolved: ResolvedDependency[];
  /** Dependencies that are already installed (by slug) */
  alreadyInstalled: string[];
  /** Dependencies not found in the index (external/unknown) */
  unresolved: string[];
}

File: src-tauri/src/models/addon.rs
use serde::{Deserialize, Serialize};

/// Represents an addon installed on the local system
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct InstalledAddon {
    pub id: i64,
    pub slug: String,
    pub name: String,
    pub installed_version: String,
    pub source_type: SourceType,
    pub source_repo: Option<String>,
    pub installed_at: String,
    pub updated_at: String,
    pub auto_update: bool,
    pub manifest_path: String,
}

/// Source type for an installed addon
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "lowercase")]
pub enum SourceType {
    Index,
    Github,
    Local,
}

impl std::fmt::Display for SourceType {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            SourceType::Index => write!(f, "index"),
            SourceType::Github => write!(f, "github"),
            SourceType::Local => write!(f, "local"),
        }
    }
}

impl std::str::FromStr for SourceType {
    type Err = String;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        match s.to_lowercase().as_str() {
            "index" => Ok(SourceType::Index),
            "github" => Ok(SourceType::Github),
            "local" => Ok(SourceType::Local),
            _ => Err(format!("Unknown source type: {}", s)),
        }
    }
}

/// Custom GitHub repository tracked by the manager
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CustomRepo {
    pub id: i64,
    pub repo: String,
    pub branch: String,
    pub release_type: ReleaseType,
    pub added_at: String,
    pub last_checked: Option<String>,
}

/// Type of release to track from GitHub
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "lowercase")]
pub enum ReleaseType {
    Release,
    Branch,
}

impl std::fmt::Display for ReleaseType {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            ReleaseType::Release => write!(f, "release"),
            ReleaseType::Branch => write!(f, "branch"),
        }
    }
}

impl std::str::FromStr for ReleaseType {
    type Err = String;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        match s.to_lowercase().as_str() {
            "release" => Ok(ReleaseType::Release),
            "branch" => Ok(ReleaseType::Branch),
            _ => Err(format!("Unknown release type: {}", s)),
        }
    }
}

/// Parsed addon manifest from .txt file
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct AddonManifest {
    pub title: String,
    pub api_version: Option<String>,
    pub author: Option<String>,
    pub version: Option<String>,
    pub description: Option<String>,
    pub dependencies: Vec<String>,
    pub optional_dependencies: Vec<String>,
    pub saved_variables: Vec<String>,
    pub files: Vec<String>,
}

/// Information about an available update
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct UpdateInfo {
    pub slug: String,
    pub name: String,
    pub current_version: String,
    pub new_version: String,
    pub download_url: String,
    /// Source type for the addon (index, github, local)
    pub source_type: SourceType,
    /// Repository for GitHub addons
    pub source_repo: Option<String>,
    /// Installation info for proper extraction (from index)
    pub install_info: Option<super::InstallInfo>,
}

/// Download progress event
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct DownloadProgress {
    pub slug: String,
    pub status: DownloadStatus,
    pub progress: f64,
    pub error: Option<String>,
}

/// Status of a download operation
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "lowercase")]
pub enum DownloadStatus {
    Pending,
    Downloading,
    Extracting,
    Complete,
    Failed,
}
